### ISSUE TITLE
Prevent or Warn when creating or updating an ILM policy using rollover with max_docs set to a value larger than the possible Lucene max docs limit

### ISSUE THREAD

**Issue by geekpete** [elastic/elasticsearch#72786]
**Elasticsearch version** (`bin/elasticsearch --version`): 7.12.1

**Plugins installed**: []

**JVM version** (`java -version`):

**OS version** (`uname -a` if on a Unix-like system):

**Description of the problem including expected versus actual behavior**:

[ILM Rollover](https://www.elastic.co/guide/en/elasticsearch/reference/master/ilm-rollover.html#ilm-rollover-options) `max_docs` option  lets you use a number larger than the Lucene max docs limit for this value and should probably instead warn or prevent you from doing this.

Related Kibana issue: https://github.com/elastic/kibana/issues/99438

**Steps to reproduce**:

Please include a *minimal* but *complete* recreation of the problem,
including (e.g.) index creation, mappings, settings, query etc.  The easier
you make for us to reproduce it, the more likely that somebody will take the
time to look at it.

 1. Try adding or updating a policy with a `max_docs` rollover option above the Lucene limit (`2147483519`)

```
PUT _ilm/policy/test-max-docs
{
  "policy": {
    "phases": {
      "hot": {
        "actions": {
          "rollover": {
            "max_age": "30d",
            "max_size": "50gb",
            "max_docs": **9147483504**
          },
          "set_priority": {
            "priority": 100
          }
        },
        "min_age": "0ms"
      }
    }
  }
}
```

**Provide logs (if relevant)**:

`[2021-05-06T14:48:43,781][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [somehost] adding index lifecycle policy [test-max-docs]`

**Comment by gaobinlong** [comment]
@geekpete , can I work on this issue?

**Comment by jakelandis** [comment]
> can I work on this issue?

I am not aware of anyone else working this. Feel free to ping me on the PR and I can help find a reviewer. Thanks !

**Comment by andreidan** [comment]
As @bhargavkn started working on this, and I initially looked at reviewing that work, it occurred to me that we should not be adding this validation at all (sorry for the late penny drop).
The rollover condition looks at `max_docs` across all the shards of an index. An elasticsearch shard is a lucene index. So an elasticsearch _index_ will be composed of possibly multiple _lucene indices_. So an elasticsearch index is able to hold more documents than a lucene index, if the elasticsearch index has multiple shards.
As an ILM policy can be configured to any index, the policy definition being disconnected from the indices it is applied to, I believe this validation should not be implemented at all.

Adding team-discuss label to have a chat about this, in case I'm missing something.

**Comment by geekpete** [comment]
Right, this totally depends on how many shards your index has so this is not really a useful or applicable limit to put into ILM. Any defence against hitting this Lucene max docs limit shouldn't be implemented in the ILM index-level context.

**Comment by bhargavkn** [comment]
@andreidan @geekpete  Yep, makes sense. I'll close the PR.

**Comment by andreidan** [comment]
We discussed this today and we indeed reached the conclusion we don't need to implement such a validation for the reasons described in https://github.com/elastic/elasticsearch/issues/72786#issuecomment-860571702 so I'm going to close this issue.

@bhargavkn I apologize that the issue was originally marked as "good first issue" and that I realised we don't need to do it only after you iterated on the PR once. We greatly appreciate your interest in contributing to Elasticsearch and I hope the experience with this issue will not put you off from contributing in the future.
